channelacceptor: add new fields to acceptor response

carlaKC opened the pull request:
This PR adds upfront shutdown, a custom error and more fields the the channel acceptor, providing clients with more control over their incoming channel parameters. The first commit is mostly a code move for the sake of unit tests, but required some changes to make it part of the acceptor struct itself (see commit message for details). 

Now that we have non-boolean fields in this response, it is more difficult to create a combined response for our chained acceptor. If we get an ambiguous set of responses (eg, two different upfront shutdown addrs), we log an error and reject the channel. I also considered erroring-out completely, but that would shutdown your channel acceptor, which isn't ideal (rather accept nothing than fail and accept errrything). 

Note on validation:
- If peer does not support upfront shutdown, funding flow fails later
- We validate proposed reserve against the dust limit sent to us by the initiating peer

To try this out I made an ugly cli tool, see last commit of [this branch](https://github.com/carlaKC/lnd/tree/4471-cliacceptor)

Fixes: #4471
Fixes: #3552

Crypt-iQ commented:
In the process of diving in, need to refresh my memory a little bit about `chanacceptor`. +1 for removing the RpcAcceptor code from the rpcserver 🔢

Crypt-iQ commented on fundingmanager.go:
```diff
@@ -1539,7 +1562,9 @@ func (f *fundingManager) handleFundingAccept(peer lnpeer.Peer,
 	// reservation.
-	chanReserve := f.cfg.RequiredRemoteChanReserve(resCtx.chanAmt, msg.DustLimit)
+	chanReserve := f.cfg.RequiredRemoteChanReserve(
+		msg.ChannelReserve, resCtx.chanAmt, msg.DustLimit,
```
I think this allows the counterparty to control their own `ChannelReserve` - the terminology between "our contribution" / "remote contribution" here always confuses me

carlaKC replied:
Updated this, decided to leave the functions alone completely and just overwrite with chan acceptor values in `handleFundingOpen` if they are present. Simpler IMO, and fewer code changes.

Crypt-iQ commented on rpcserver.go:
```diff
@@ -6363,7 +6363,13 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
-		stream.Recv, stream.Send, r.cfg.AcceptorTimeout, r.quit,
+		stream.Recv, stream.Send, r.cfg.AcceptorTimeout,
+		func(address string) (lnwire.DeliveryAddress, error) {
+			return parseUpfrontShutdownAddress(
```
non-blocking:
Might be better to export this function and then pass in `ActiveNetParams` to the `RPCAcceptor`

carlaKC replied:
Problem here is that we can't import something in `rpcserver` into the `chanacceptor` package because we get a circular dep. Could move it and export, but I couldn't pick a logical place for it because it's used in a few places in the server.

Crypt-iQ replied:
Maybe the `chancloser` package?

Crypt-iQ reviewed (commented):
Light first pass review done - still need to test locally

alexbosworth commented on lnrpc/rpc.proto:
```diff
@@ -791,6 +791,52 @@ message ChannelAcceptResponse {
+    The reserve amount in satoshis that we require the remote peer to adhere to.
+    We require that the remote peer always have some reserve amount allocated to
+    them so that there is always a disincentive to broadcast old state (if they
+    hold 0 sats on their side of the channel, there is nothing to loose).
```
```suggestion
    hold 0 sats on their side of the channel, there is nothing to lose).
```

carlaKC replied:
tq!

carlaKC commented on lnrpc/rpc.proto:
```diff
@@ -791,6 +791,52 @@ message ChannelAcceptResponse {
+    false, the channel will be rejected if an error is set with accept=true
+    because the meaning of this response is ambiguous.
+    */
+    string error = 3;
```
Good idea, adding one! Any suggestions re the limit amount? 500 chars?

Crypt-iQ replied:
Yeah that seems sane

carlaKC commented:
Updated to add a few more fields to the response (CSV delay + confirmations required), and a length limit on custom errors as suggested. Also simplified the funding manager changes a bit. Can be tested out with `lncli acceptchans` from this [branch](https://github.com/carlaKC/lnd/tree/4471-cliacceptor).

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -1,24 +1,227 @@
+	// listening and any in-progress requests should be terminated.
+	defer close(r.done)
+
+	// Create a channel that responses
```
?

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -1,24 +1,227 @@
+
+	newRequest := &chanAcceptInfo{
+		Request:  req,
+		Response: respChan,
```
Don't need to be exported

Crypt-iQ reviewed (commented):
ACK fff9ee4cc98782c17059dfaf0b09ecc02d66ec17

Still looking at the other 2 commits

carlaKC commented on chanacceptor/acceptor_test.go:
```diff
@@ -1,17 +1,18 @@
 	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/lnrpc"
 	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/stretchr/testify/assert"
```
there used to be an issue in testify where require would have some strange behaviour if you used it in a goroutine, will check whether that's still the case.

Crypt-iQ commented on chanacceptor/acceptor_test.go:
```diff
@@ -210,3 +213,41 @@ func TestMultipleAcceptClients(t *testing.T) {
+
+// TestInvalidResponse tests the case where our remote channel acceptor sends us
+// an invalid response, so the channel acceptor stream terminates.
+func TestInvalidResponse(t *testing.T) {
```
I like this test 🌠

Crypt-iQ commented on fundingmanager.go:
```diff
@@ -1274,10 +1276,13 @@ func (f *fundingManager) handleFundingOpen(peer lnpeer.Peer,
-	if !f.cfg.OpenChannelPredicate.Accept(chanReq) {
+	// Query our channel acceptor to determine whether we should reject
+	// the channel.
+	acceptorResp := f.cfg.OpenChannelPredicate.Accept(chanReq)
```
Very nice, minimal change to the fundingmanager !

Crypt-iQ commented on chanacceptor/acceptor_test.go:
```diff
@@ -251,3 +277,46 @@ func TestInvalidResponse(t *testing.T) {
+		reserve   = dustLimit / 2
+
+		// We make a single query, and expect it to fail with our
+		// generic error because channel reserve it too low.
```
nit: it -> is

Crypt-iQ reviewed (commented):
Another pass done - I'll probably want to get another look after somebody else gives a review. A benefit of the new code structuring is that we could probably make an RPC call for adjusting the timeout of an individual acceptor. This may require some map or some memory in the `rpcserver` but it is possible and may even be desirable. Of course perhaps a *new* rpcacceptor could be made instead and you just adjust a global timeout that only affects new acceptors. I know that certain folks have different time preferences like mobile users may want longer timeouts to perhaps alert the user with a push notification that an open channel is incoming. Automated scripts would need a lower timeout.

carlaKC commented:
Just addresses some nits in [last push](https://github.com/lightningnetwork/lnd/compare/8d90348495fda9825d184d82b16ceddf30ecb0cc..ec21c7b8af3365c2469d6ea85c486958549c0b69).

bhandras reviewed (commented):
I think the PR looks really good. My comments are mostly nits and comments. :taco:

carlaKC commented on chanacceptor/acceptor_test.go:
```diff
@@ -1,156 +1,206 @@
-		}
+	case <-time.After(timeout):
+		c.t.Fatalf("timeout receiving request")
+		return nil, fmt.Errorf("timeout")
```
won't compile without it?

carlaKC commented on chanacceptor/rpcacceptor.go:
```diff
@@ -255,21 +282,41 @@ func (r *RPCAcceptor) Run() error {
+	// Ensure that the reserve that has been proposed, if it is set, is at
+	// least the dust limit that was proposed by the remote peer.
+	reserveSat := btcutil.Amount(req.ReserveSat)
+	if reserveSat != 0 && reserveSat < dustLimit {
```
BOLT 2 says `MUST set channel_reserve_satoshis greater than or equal to dust_limit_satoshis`
The error message here is misleading though, changing "greater than" -> "at least" to make it clearer

Crypt-iQ replied:
Should this have a comment that this is Bolt002 - required?

carlaKC commented on chanacceptor/chainedacceptor.go:
```diff
@@ -61,11 +67,135 @@ func (c *ChainedAcceptor) Accept(req *ChannelAcceptRequest) *ChannelAcceptRespon
+	}
+	current.MinHtlcIn = val.(lnwire.MilliSatoshi)
+
+	val, err = mergeValue(
```
I'd rather keep it strict and require the external acceptors to provide us with a consistent response than try to infer meaning from the fields provided?

I think that most people's use case is covered by a single acceptor (don't need to worry about this). And in the case of multiple accceptors (say you have different services running different acceptors - one for upfront shutdown, one for rate limiting etc), it seems strange that they'd be setting the same fields if they have this level of specialization?

bhandras replied:
Yeah, I think your reasoning makes perfect sense.

carlaKC commented on chanacceptor/interface.go:
```diff
@@ -33,18 +34,58 @@ type ChannelAcceptResponse struct {
+
+	// UpfrontShutdown is the address that we will set as our upfront
+	// shutdown address if the peer supports the feature bit.
+	UpfrontShutdown lnwire.DeliveryAddress
```
Generally a bit unsure of this type of comment, because it's easy for the merge function to up renamed/changed without the comment being updated, and ideally people who add new fields will unit test? Not strong feelings, so added it, but more of a meta thought.

bhandras replied:
I think commenting probably doesn't hurt in this case as the current unit test would miss added fields.

carlaKC commented on chanacceptor/rpcacceptor.go:
```diff
@@ -1,24 +1,227 @@
+		for {
+			resp, err := r.receive()
+			if err != nil {
+				errChan <- err
```
good point, done!

bhandras reviewed (approved):
LGTM :waffle:

cfromknecht commented on chanacceptor/rpcacceptor.go:
```diff
@@ -232,9 +246,27 @@ func (r *RPCAcceptor) sendAcceptRequests(errChan chan error,
+
+			// Validate the response we have received. If it is not
+			// valid, we log our error and reject the channel.
+			if err := validateAcceptorResponse(resp); err != nil {
```
style nit: perhaps it would simplify this block to have `validateAcceptorResponse` return the final `accept` and `acceptErr` values? something like:
```golang
accept, acceptErr, err := validateAcceptorResponse(resp)
if err != nil {
     log...
}
```

cfromknecht commented on chanacceptor/chainedacceptor.go:
```diff
@@ -61,11 +67,135 @@ func (c *ChainedAcceptor) Accept(req *ChannelAcceptRequest) *ChannelAcceptRespon
+
+	// If both values are not empty, but are not equal, then we have an
+	// inconsistent set of values.
+	case current != new:
```
not positive, but my intuition says that direct interface comparisons could have some unexpected behavior. for example, what happens when merging two non-nil `UpfrontShutdownScripts`? in this playground example, attempting to compare two interfaced byte slices causes a runtime panic https://play.golang.org/p/cC53YdpsCwW. maybe i'm missing something that prevents this?

carlaKC replied:
TIL, great catch! Updated to just use separate merge functions for the different types, sharing where possible. Also added tests for the merge function so that everything is covered.

cfromknecht reviewed (commented):
nice pr! looks pretty much ready, only one main comment about interfaced comparisons during merges

Crypt-iQ commented on chanacceptor/merge.go:
```diff
@@ -0,0 +1,152 @@
+
+// mergeError returns a merge error for a named field when we get two channel
+// acceptor responses which have different values set.
+func mergeError(name string, current, new interface{}) error {
```
Confusingly named because it doesnt merge two `error` types

carlaKC replied:
renamed!

Crypt-iQ commented on chanacceptor/rpcacceptor.go:
```diff
@@ -288,12 +325,36 @@ func (r *RPCAcceptor) sendAcceptRequests(errChan chan error,
 // error to send to the peer, and any validation errors that occurred.
-func validateAcceptorResponse(req lnrpc.ChannelAcceptResponse) (bool, error,
-	error) {
+func (r *RPCAcceptor) validateAcceptorResponse(dustLimit btcutil.Amount,
```
Is there anything else we'd like to validate against here?

carlaKC replied:
All covered in terms of the requirements for BOLT2. Also added a check that we don't set > 483 for max htlcs. The other side should catch it anyway, but might as well validate early.

Crypt-iQ reviewed (commented):
Looks good - two minor comments. If the merge logic gets larger, it might be good to have quickcheck tests to sanity check a lot of different merge combos.

carlaKC commented:
> If the merge logic gets larger

I think we should be good since we have all the fields in the `AcceptChannel` message except for the basepoints.

Crypt-iQ reviewed (approved):
LGTM 🐱‍👤

cfromknecht commented on chanacceptor/rpcacceptor_test.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// TestValidateAcceptorResponse test validation of acceptor responses.
+func TestValidateAcceptorResponse(t *testing.T) {
```
nice test :)

cfromknecht commented on chanacceptor/chainedacceptor.go:
```diff
@@ -61,11 +63,31 @@ func (c *ChainedAcceptor) Accept(req *ChannelAcceptRequest) *ChannelAcceptRespon
 			return acceptorResponse
 		}
+
+		// If we have accepted the channel, we need to set the other
```
great comment

cfromknecht commented on chanacceptor/merge.go:
```diff
@@ -0,0 +1,152 @@
+// mergeResponse takes two channel accept responses, and attempts to merge their
+// fields, failing if any fields conflict (are non-zero and not equal). It
+// returns a new response that has all the merged fields in it.
+func mergeResponse(current, new ChannelAcceptResponse) (ChannelAcceptResponse,
```
dig the additional type safety, forces us to consider (and test) all possible combos :)

cfromknecht commented on chanacceptor/merge_test.go:
```diff
@@ -0,0 +1,188 @@
+)
+
+// TestMergeResponse tests merging of channel acceptor responses.
+func TestMergeResponse(t *testing.T) {
```
🥇

cfromknecht commented on chanacceptor/chainedacceptor.go:
```diff
@@ -61,11 +63,31 @@ func (c *ChainedAcceptor) Accept(req *ChannelAcceptRequest) *ChannelAcceptRespon
+			log.Errorf("response for: %x has inconsistent values: %v",
+				req.OpenChanMsg.PendingChannelID, err)
+
+			return NewChannelAcceptResponse(
```
note to self: if we ever do this in more than one place, define a specialized constructor

cfromknecht reviewed (approved):
LGTM! 🥂
